imputation/ts_impute.go

```diff
@@ -50,42 +59,102 @@ func dateGapFinder(keys []string) (string, int, int, int) {
 
 	year, month, day := 0, 0, 0
 
-	duration := 1<<31 - 1 // A large number.
+	equalMonths := true
+	equalDays := true
+
```

Comment:
It took over an hour to think through the algorithm and code, not because you wrote poor code but because I had trouble thinking through how to write the code succinctly but efficiently.  The main problem (I think) is that we have to loop through the data to determine which GCD to compute and then we can compute it.

The algorithm is fairly easy to state if we had vector operations such as

- allEqual which returns true iff all the values are the same,
- \Delta which computes the differences of adjacent values,
- gcd applied to a vector.

We do not have those.

My struggles yielded https://play.golang.org/p/VHeCA9GtZmI and https://play.golang.org/p/hPioJ8xPVl6.

I do not know what the best approach is.  Most of the time the inputs are parsed twice, first by strings and then as a Time.  Another approach would be to parse once as a Time, storing the parsed values.  Another approach would be to parse once as a Time and compute all possible values (year gcd, month gcd, day gcd, all months equal, and all days equal) and then return the proper value.  In the best of all worlds, I would like to write the vectorized operations and have the optimizer convert the code into more efficient code.  We do not live in that world.

The latter example keeps running totals as it runs through the data just as the code here does.  I find running totals confusing, but it does avoid parsing the inputs twice.  However, it computes three GCDs when only one may be necessary.

I think your code may be the most efficient.